library(openxlsx)
library(tidyverse)
library(janitor)
source("Excel_Format_Params.r")

# @sheets: all sheets from the excel generated by 06_01_Cofdr_Format_Excel.r
format_sig_snp <- function(pheno, sheets)
{
	print(paste(pheno, "start!"))
	dat = sheets[[which(names(sheets)==pheno)]]
	
	# T1.sig_snps
	if(is.null(dat$T1.sig_snps)){
		sig_snps = as.data.frame(matrix(ncol=2, nrow=0))
	} else{
		sig_snps = dat$T1.sig_snps[-1, ] %>% row_to_names(row_number = 1) %>% select(SNP)
		sig_snps = as.data.frame(cbind(sig_snps, pheno))
	}	
	colnames(sig_snps) = c("SNP", "comparisons")
	
	# T2.risk_loci
	if(is.null(dat$T2.risk_loci)){
		risk_loci = as.data.frame(matrix(ncol=2, nrow=0))
	} else{
		risk_loci = dat$T2.risk_loci[-1, ] %>% row_to_names(row_number = 1) %>% select(rsID)
		risk_loci = as.data.frame(cbind(risk_loci, pheno))	
	}
	colnames(risk_loci) = c("SNP", "comparisons")
		
	# T7.commonLoci_cofdr_hypr
	if(is.null(dat$T7.commonLoci_cofdr_hypr)){
		commonLoci_cofdr_hypr = as.data.frame(matrix(ncol=2, nrow=0))
	} else{
		commonLoci_cofdr_hypr = dat$T7.commonLoci_cofdr_hypr[-1, ] %>% row_to_names(row_number = 1) %>% filter(total_detect==2) %>% select(SNP)
		if(nrow(commonLoci_cofdr_hypr)>0){
			commonLoci_cofdr_hypr = as.data.frame(cbind(commonLoci_cofdr_hypr, pheno))		
		} else{
			commonLoci_cofdr_hypr = as.data.frame(matrix(ncol=2, nrow=0))		
		}	
	}	
	colnames(commonLoci_cofdr_hypr) = c("SNP", "comparisons")
	
	# T8.gene_based_cofdr (total_detect, magma_detect, twas_detect, sptwas_detect, pwas_detect)
	if(is.null(dat$T8.gene_based_cofdr)){
		gene_based_cofdr = mgene = egene = sgene = pgene = as.data.frame(matrix(ncol=2, nrow=0))
	} else{
		gene_based_cofdr = dat$T8.gene_based_cofdr[-1, ] %>% row_to_names(row_number = 1) %>% filter(total_detect>=3) %>% select(hgnc_symbol)
		mgene = dat$T8.gene_based_cofdr[-1, ] %>% row_to_names(row_number = 1) %>% filter(magma_detect==1) %>% select(hgnc_symbol)
		egene = dat$T8.gene_based_cofdr[-1, ] %>% row_to_names(row_number = 1) %>% filter(twas_detect==1) %>% select(hgnc_symbol)
		sgene = dat$T8.gene_based_cofdr[-1, ] %>% row_to_names(row_number = 1) %>% filter(sptwas_detect==1) %>% select(hgnc_symbol)
		pgene = dat$T8.gene_based_cofdr[-1, ] %>% row_to_names(row_number = 1) %>% filter(pwas_detect==1) %>% select(hgnc_symbol)
		
		format_gene <- function(gene){
			if(nrow(gene)>0){
				gene = as.data.frame(cbind(unique(gene), pheno))		
			} else{
				gene = as.data.frame(matrix(ncol=2, nrow=0))		
			}
			return(gene)
		}
		
		gene_based_cofdr = format_gene(gene_based_cofdr)
		mgene = format_gene(mgene)
		egene = format_gene(egene)
		sgene = format_gene(sgene)
		pgene = format_gene(pgene)
	}	
	colnames(gene_based_cofdr) = colnames(mgene) = colnames(egene) = colnames(sgene) = colnames(pgene) = c("Gene", "comparisons")
	
	return(list(sig_snps, risk_loci, commonLoci_cofdr_hypr, gene_based_cofdr, mgene, egene, sgene, pgene))
}

Multi_Cofdr_Excel_Summary <- function(excel_dir, save_path)
{
	paths = list.files(excel_dir, full.names=TRUE)	
	phenos = stringi::stri_replace_all_regex(basename(paths), pattern=c("_0.8TT.xlsx", "_no_sigsnp.xlsx", ".xlsx"), replacement=c("", "", ""), vectorize=FALSE)

	sheets = pbapply::pblapply(paths, read_all_sheets)	
	names(sheets) = phenos
	res = lapply(phenos, format_sig_snp, sheets=sheets)
	
	# idx: index for "sig_snps, risk_loci, commonLoci_cofdr_hypr, gene_based_cofdr, sig_gene_GSEA"
	# col1: name of the 1st column
	format_Excel <- function(idx, col1)
	{
		print(idx)
		dat = lapply(res, '[[', idx) 
		dat = dat[lapply(dat, nrow)>0] 
		
		if(!identical(dat, list())){
			dat = dat %>% reduce(full_join, by=col1)
			if(ncol(dat)==2){
				dat$N_comparisons = 1
			}else{
				dat$N_comparisons = apply(dat[, -1], 1, function(x) sum(!is.na(x)))
				dat$comparisons = apply(dat[, -c(1,ncol(dat))], 1, paste, collapse = ";")
			}
			dat = dat[, c(col1, "N_comparisons", "comparisons")]
			dat = dat[order(dat$N_comparisons, decreasing=TRUE),]
		} else{
			dat = setNames(as.data.frame(matrix(ncol=3, nrow=1)), c(col1, "N_comparisons", "comparisons"))
		}
			return(dat)
	}

	wb <- createWorkbook()
	addWorksheet(wb, "T1.sig_snps")
	addWorksheet(wb, "T2.risk_loci")
	addWorksheet(wb, "T3.commonLoci_cofdr_hypr")
	addWorksheet(wb, "T4.Validated.Gene")
	addWorksheet(wb, "T5.Sig.Gene.MAGMA")
	addWorksheet(wb, "T6.Sig.Gene.TWAS")
	addWorksheet(wb, "T7.Sig.Gene.spTWAS")
	addWorksheet(wb, "T8.Sig.Gene.PWAS")

	h1 = "Table 1. Significantly colocalized SNPs detected by co-fdr at posterior possibility threshold 0.9 across all multiple trait comparisons"
	subh1 = 'SNP=rsid of the colocalized SNP; N_comparisons=number of comparisons identifying the "SNP"; comparisons=name of trait comparisons identifying the "SNP"'
	dat1 = format_Excel(1, "SNP")
	format_header(wb, "T1.sig_snps", c(1, 12), h1, header_style)
	format_header(wb, "T1.sig_snps", c(2, 12), subh1, subheader_style)
	format_main(wb, "T1.sig_snps", dat1, title_style, main_style)
	
	h2 = "Table 2. Identification of genomic risk loci across all multiple trait comparisons"
	subh2 = 'SNP=rsid of the lead SNP in the genomic locus; N_comparisons=number of comparisons identifying the "SNP"; comparisons=name of trait comparisons identifying the "SNP"'
	dat2 = format_Excel(2, "SNP")
	format_header(wb, "T2.risk_loci", c(1, 13), h2, header_style)
	format_header(wb, "T2.risk_loci", c(2, 13), subh2, subheader_style)
	format_main(wb, "T2.risk_loci", dat2, title_style, main_style)
	
	h3 = "Table 3. Significantly colocalized SNPs detected by cofdr and hyprcoloc across all multiple trait comparisons"
	subh3 = 'SNP=rsid of the colocalized SNP; N_comparisons=number of comparisons identifying the "SNP"; comparisons=name of trait comparisons identifying the "SNP"'
	dat3 = format_Excel(3, "SNP")
	format_header(wb, "T3.commonLoci_cofdr_hypr", c(1, 12), h3, header_style)
	format_header(wb, "T3.commonLoci_cofdr_hypr", c(2, 12), subh3, subheader_style)
	format_main(wb, "T3.commonLoci_cofdr_hypr", dat3, title_style, main_style)
	
	h4 = "Table 4. Gene-based co-fdr analysis to detect genes associated with both traits using MAGMA, PrediXcan or FUSION across all multiple trait comparisons"
	subh4 = 'Gene=HGNC symbol of the colocalized gene detected by at least three pathways of MAGMA, TWAS, spTWAS, and PWAS; N_comparisons=number of comparisons identifying the "Gene"; comparisons=name of trait comparisons identifying the "Gene"'
	dat4 = format_Excel(4, "Gene")
	format_header(wb, "T4.Validated.Gene", c(1, 14), h4, header_style)
	format_header(wb, "T4.Validated.Gene", c(2, 14), subh4, subheader_style)
	format_main(wb, "T4.Validated.Gene", dat4, title_style, main_style)
	
	h5 = "Table 5. Gene-based co-fdr analysis to detect genes associated with both traits using MAGMA results only across all multiple trait comparisons"
	subh5 = 'Gene=HGNC symbol of the colocalized gene detected by MAGMA; N_comparisons=number of comparisons identifying the "Gene"; comparisons=name of trait comparisons identifying the "Gene"'
	dat5 = format_Excel(5, "Gene")
	format_header(wb, "T5.Sig.Gene.MAGMA", c(1, 15), h5, header_style)
	format_header(wb, "T5.Sig.Gene.MAGMA", c(2, 15), subh5, subheader_style)
	format_main(wb, "T5.Sig.Gene.MAGMA", dat5, title_style, main_style)
	
	h6 = "Table 6. Gene-based co-fdr analysis to detect genes associated with both traits using TWAS results only across all multiple trait comparisons"
	subh6 = 'Gene=HGNC symbol of the colocalized gene detected by TWAS; N_comparisons=number of comparisons identifying the "Gene"; comparisons=name of trait comparisons identifying the "Gene"'
	dat6 = format_Excel(6, "Gene")
	format_header(wb, "T6.Sig.Gene.TWAS", c(1, 15), h6, header_style)
	format_header(wb, "T6.Sig.Gene.TWAS", c(2, 15), subh6, subheader_style)
	format_main(wb, "T6.Sig.Gene.TWAS", dat6, title_style, main_style)
	
	h7 = "Table 7. Gene-based co-fdr analysis to detect genes associated with both traits using spTWAS results only across all multiple trait comparisons"
	subh7 = 'Gene=HGNC symbol of the colocalized gene detected by spTWAS; N_comparisons=number of comparisons identifying the "Gene"; comparisons=name of trait comparisons identifying the "Gene"'
	dat7 = format_Excel(7, "Gene")
	format_header(wb, "T7.Sig.Gene.spTWAS", c(1, 15), h7, header_style)
	format_header(wb, "T7.Sig.Gene.spTWAS", c(2, 15), subh7, subheader_style)
	format_main(wb, "T7.Sig.Gene.spTWAS", dat7, title_style, main_style)
	
	h8 = "Table 8. Gene-based co-fdr analysis to detect genes associated with both traits using PWAS results only across all multiple trait comparisons"
	subh8 = 'Gene=HGNC symbol of the colocalized gene detected by PWAS; N_comparisons=number of comparisons identifying the "Gene"; comparisons=name of trait comparisons identifying the "Gene"'
	dat8 = format_Excel(8, "Gene")
	format_header(wb, "T8.Sig.Gene.PWAS", c(1, 15), h8, header_style)
	format_header(wb, "T8.Sig.Gene.PWAS", c(2, 15), subh8, subheader_style)
	format_main(wb, "T8.Sig.Gene.PWAS", dat8, title_style, main_style)
	
	saveWorkbook(wb, file = save_path, overwrite = TRUE)
	
	return(save_path)
}


# Multi_Cofdr_Excel_Summary(excel_dir=paste0("03_cofdr/11_Multi_Cofdr_Summary/Tables/", c("A2", "B2", "C2")), save_path=paste0("03_cofdr/11_Multi_Cofdr_Summary/Tables/A2_B2_C2_respiratory.xlsx"))