# please modify paths of required files/tools or other paramaters before any analysis 

# 1. tools to be installed 
# plink 
plink_bin = "/usr/local/bin/plink" # https://www.cog-genomics.org/plink/

# MAGMA
old_path = Sys.getenv("PATH")
Sys.setenv(PATH = paste(old_path, "/mnt/data/xue/Tool/MAGMA", sep=":")) # https://ctg.cncr.nl/software/magma

# GCTA
gcta_bin = "/mnt/data/xue/Tool/gcta-1.94.1-linux-kernel-3-x86_64/gcta-1.94.1" # https://yanglab.westlake.edu.cn/software/gcta/#Download

# gwas-pw
old_path = Sys.getenv("PATH")
Sys.setenv(PATH = paste(old_path, "/mnt/data/xue/Tool/gwas-pw-0.21/src", sep=":")) # https://github.com/joepickrell/gwas-pw
	
# 2. files to be downloaded 
# clumping
clump_kb = 1000
clump_r2 = 0.05
clump_p = 1
pop = "EUR"
bfile = paste0("/mnt/data/xue/Tool/MAGMA/reference/", tolower(pop), "/g1000_", tolower(pop)) # https://ctg.cncr.nl/software/MAGMA/ref_data/g1000_eur.zip

# MAGMA Entrze id 
upstream_kb = 0
downstream_kb = 0
id_type = "Entrze"; genomeLocFile = "/mnt/data/xue/Tool/MAGMA/auxiliary/NCBI37.3.gene.loc" # https://ctg.cncr.nl/software/MAGMA/aux_files/NCBI37.3.zip
genome_ref_path = bfile
genesAnnotPrefix = paste0(genome_ref_path, ".", upstream_kb, "UP.", downstream_kb, "DOWN.", id_type)
set_name = "msigdb"; set_path = "/mnt/data/xue/Tool/MAGMA/set/MSigDB/c2c5.v7.5.1.entrez.gmt" # https://www.gsea-msigdb.org/gsea/msigdb/download_file.jsp?filePath=/msigdb/release/2023.2.Hs/c2.all.v2023.2.Hs.entrez.gmt; https://www.gsea-msigdb.org/gsea/msigdb/download_file.jsp?filePath=/msigdb/release/2023.2.Hs/c5.all.v2023.2.Hs.entrez.gmt

# MAGMA ENSG id (Ignore this if performing MAGMA analysis in FUMA)
id_type_ensg = "ENSG"; genomeLocFile_ensg = "/mnt/data/xue/Tool/FUMA/auxiliary/GRCh37.v92.ENSG.protein_coding.genes.txt" # extract protein coding genes from https://github.com/Kyoko-wtnb/FUMA_scRNA_data/blob/master/ENSG.genes.txt
genesAnnotPrefix_ensg = paste0(genome_ref_path, ".", upstream_kb, "UP.", downstream_kb, "DOWN.", id_type_ensg)
scRNA_paths = list.files(pattern = "*.txt.gz", path = "/mnt/data/xue/Tool/FUMA/scRNA", full.names = TRUE) # https://github.com/Kyoko-wtnb/FUMA_scRNA_data
scRNA_names = as.character(strsplit(basename(scRNA_paths), ".txt.gz"))
covar_name = c("gtex_v8_ts", "gtex_v8_ts_general", scRNA_names)
covar_path = c("/mnt/data/xue/Tool/FUMA/GTEx_v8_RNAseq/gtex_v8_ts_avg_log2TPM.txt", "/mnt/data/xue/Tool/FUMA/GTEx_v8_RNAseq/gtex_v8_ts_general_avg_log2TPM.txt", scRNA_paths) # requested from FUMA

# PrediXcan models with tissues as 1st column
model_src = read.table("05_auxiliary/predixcan_masher_models.txt", header = FALSE) 

# sqtl annotation: map introns to genes (You may also download the annotation file generated by PrediXcan https://uchicago.app.box.com/s/xy71r0su6refrfggivrwc7kpvsxysmmn)
sqtl_ann = "05_auxiliary/gencode_hg38_to_hg19_all_introns.ann.bed.gz" # We extracted and reformatted annotation for excision of intron juctions only https://github.com/davidaknowles/leafcutter/issues/67  

# gene annotation
gene_annot = "/mnt/data/xue/Tool/FUMA/auxiliary/GRCh37.v92.ENSG.genes.txt" # https://github.com/Kyoko-wtnb/FUMA_scRNA_data/blob/master/ENSG.genes.txt

# ld score
ld_chr = paste0("/mnt/data/xue/Tool/ldsc/ld_ref/", tolower(pop), "_w_ld_chr/") # https://github.com/bulik/ldsc

# ld block
ldetect_data = paste0("/mnt/data/xue/Tool/gwas-pw-0.21/ldetect_data/fourier_ls-all_", pop, ".bed") # https://bitbucket.org/nygcresearch/ldetect-data/src/master/

# 3. other files to be prepared 
# sample size
sample_size_path = "01_data/sample_size.txt"
sample_size_src = read.table(sample_size_path, row.names = 1, header = TRUE, sep = "\t") # traits as rownames and "pheno N_CAS N_CON N N_EFF" as colnames

# LDSC, generated by 04_script/
ldsc_obs_h2_path = "01_data/01_LDSC_Results/ldsc_h2.txt"
ldsc_rg_path = "01_data/01_LDSC_Results/ldsc_rg.txt"
ldsc_intercept_path = "01_data/01_LDSC_Results/ldsc_intercept.txt" # use tmvtnorm estimates "01_data/02_tmvtnorm_Results/tmvtnorm_overlapCor.txt" instead when LDSC estimates are "NA" due to limited sample size
ldsc_obs_h2_src = read.table(ldsc_obs_h2_path, row.names = 1, header = TRUE, sep = "\t")
ldsc_rg_src = read.table(ldsc_rg_path, row.names = 1, header = TRUE, sep = "\t")
ldsc_intercept_src = read.table(ldsc_intercept_path, row.names = 1, header = TRUE, sep = "\t")




